runtime/gc_glue_java/MetronomeDelegate.hpp

```diff
@@ -237,13 +238,29 @@ class MM_MetronomeDelegate : public MM_BaseNonVirtual
 	MMINLINE UDATA
 	scanPointerRange(MM_EnvironmentRealtime *env, fj9object_t *startScanPtr, fj9object_t *endScanPtr)
 	{
-		fj9object_t *scanPtr = startScanPtr;
-		UDATA pointerFieldBytes = (UDATA)(endScanPtr - scanPtr);
-		UDATA pointerField = pointerFieldBytes / sizeof(fj9object_t);
-		while(scanPtr < endScanPtr) {
-			GC_SlotObject slotObject(_javaVM->omrVM, scanPtr);
-			_markingScheme->markObject(env, slotObject.readReferenceFromSlot());
-			scanPtr++;
+		UDATA pointerFieldBytes = 0;
+		UDATA pointerField = 0;
+	
+		if (env->compressObjectReferences()) {
+			uint32_t *scanPtr = (uint32_t*)startScanPtr;
+			uint32_t *endPtr = (uint32_t*)endScanPtr;
+			pointerFieldBytes = (UDATA)(endPtr - scanPtr);
+			pointerField = pointerFieldBytes / sizeof(uint32_t);
+			while(scanPtr < endPtr) {
+				GC_SlotObject slotObject(_javaVM->omrVM, (fj9object_t*)scanPtr);
+				_markingScheme->markObject(env, slotObject.readReferenceFromSlot());
+				scanPtr++;
+			}
+		} else {
+			uintptr_t *scanPtr = (uintptr_t*)startScanPtr;
+			uintptr_t *endPtr = (uintptr_t*)endScanPtr;
+			pointerFieldBytes = (UDATA)(endPtr - scanPtr);
+			pointerField = pointerFieldBytes / sizeof(uintptr_t);
+			while(scanPtr < endPtr) {
+				GC_SlotObject slotObject(_javaVM->omrVM, (fj9object_t*)scanPtr);
+				_markingScheme->markObject(env, slotObject.readReferenceFromSlot());
+				scanPtr++;
+			}
```

Comment:
This pass is somewhat perf critical and extra runtime checks in readReferenceFromSlot() could be tangible (sparsly populated arrays). Would it make sense to have specialized GC_SlotObject::readCRReferenceFromSlot/readNonCRReferenceFromSlot APIs?